chain/tron/txinput/tx_input.go

```diff
@@ -0,0 +1,148 @@
+package txinput
+
+import (
+	"time"
+
+	xc "github.com/cordialsys/crosschain"
+	"github.com/cordialsys/crosschain/chain/tron/core"
+	"github.com/cordialsys/crosschain/chain/tron/http_client"
+	"github.com/cordialsys/crosschain/factory/drivers/registry"
+)
+
+const TX_TIMEOUT = 2 * time.Hour
+
+// TxInput for Tron
+type TxInput struct {
+	xc.TxInputEnvelope
+
+	// 6th to 8th (exclusive) byte of the reference block height
+	RefBlockBytes []byte `json:"ref_block_bytes,omitempty"`
+	// 8th to 16th (exclusive) byte of the reference block hash
+	RefBlockHash []byte `json:"ref_block_hash,omitempty"`
+
+	// Expiration time (seconds)
+	Expiration int64 `json:"expiration,omitempty"`
+	// Transaction creation time (seconds)
+	Timestamp int64 `json:"timestamp,omitempty"`
+	// Max fee budget
+	MaxFee xc.AmountBlockchain `json:"max_fee,omitempty"`
+}
+
+var _ xc.TxInput = &TxInput{}
+var _ xc.TxInputWithUnix = &TxInput{}
+
+func init() {
+	registry.RegisterTxBaseInput(&TxInput{})
+	registry.RegisterTxVariantInput(&StakeInput{})
+	registry.RegisterTxVariantInput(&UnstakeInput{})
+	registry.RegisterTxVariantInput(&WithdrawInput{})
+}
+
+func NewTxInput() *TxInput {
+	return &TxInput{
+		TxInputEnvelope: xc.TxInputEnvelope{
+			Type: xc.DriverTron,
+		},
+	}
+}
+
+func (input *TxInput) GetDriver() xc.Driver {
+	return xc.DriverTron
+}
+
+func (input *TxInput) SetGasFeePriority(other xc.GasFeePriority) error {
+	multiplier, err := other.GetDefault()
+	if err != nil {
+		return err
+	}
+	// tron doesn't do prioritization
+	_ = multiplier
+	return nil
+}
+
+func (input *TxInput) GetFeeLimit() (xc.AmountBlockchain, xc.ContractAddress) {
+	return input.MaxFee, ""
+}
+
+func (input *TxInput) SetUnix(unix int64) {
+	input.Timestamp = unix
+	input.Expiration = unix + int64((TX_TIMEOUT).Seconds())
+}
+func (input *TxInput) IndependentOf(other xc.TxInput) (independent bool) {
+	// tron uses recent-block-hash like mechanism like solana, but with explicit timestamps
+	return true
+}
+func (input *TxInput) SafeFromDoubleSend(other xc.TxInput) (safe bool) {
+	oldInput, ok := other.(*TxInput)
+	if ok {
+		if input.Timestamp <= oldInput.Expiration {
+			return false
+		}
+	} else {
+		// can't tell (this shouldn't happen) - default false
+		return false
+	}
+	// all others timed out - we're safe
+	return true
+}
```

Comment:
this will always fail when comparing inputs of different transaction types (e.g. transfer vs staking).

Please refactor to use an interface instead that you can implement just on the base input type.

See examples in recent PR here:
https://github.com/cordialsys/crosschain/pull/174